### PR TITLE
[BD-21] Add setting toggles

### PR DESCRIPTION
**Pull request by regisb**
**Description:** This adds a `toggles` module to the edx_toggles package. We add there
classes to store feature toggles in django settings.

These classes expose an API that is similar to those of edx-platform's
waffle_utils, such that they can be added there without too many
changes.

**JIRA:** https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943

**Dependencies:** a PR will have to be created in edx-platform to make use of this new feature

**Merge deadline:** List merge deadline (if any)

**Testing instructions:**

    make test-django

**Reviewers:**
- [ ] @robrap 
- [ ] @feanil 

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @regisb! I've created [BLENDED-596](https://openedx.atlassian.net/browse/BLENDED-596) to keep track of it in Jira.   More details are on the [BD-21](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation) project page.

When this pull request is ready, tag your edX technical lead.

**Comment by regisb** [comment]
This is ready for review cc @robrap. It's currently labeled as a WIP, as it's lacking some docs, but I'd like to get a first approval before moving further.

File: edx_toggles/tests/test_toggles.py
Comment (robrap) [inline edx_toggles/tests/test_toggles.py:30]: I like the `toggle1` and `toggle2` tests above for completeness with testing `False`.

File: Makefile
Comment (robrap) [inline Makefile:67]: In the django-app cookiecutter, it just uses pytest to test everything.  See https://github.com/edx/edx-cookiecutters/blob/master/cookiecutter-django-app/%7B%7Bcookiecutter.repo_name%7D%7D/Makefile#L60-L61  Can this use that pattern?

File: edx_toggles/tests/test_toggles.py
Comment (robrap) [inline edx_toggles/tests/test_toggles.py:34]: For completeness, you could add a "key2" and test the setting (dict) being set, but the key not being set.

**Review by robrap — COMMENTED** [review]
Thank you.

File: edx_toggles/toggles.py
Comment (regisb) [reply edx_toggles/toggles.py:7]: No, I expect that the toggles from waffle_utils will inherit from this class as well.

File: edx_toggles/toggles.py
Comment (regisb) [reply edx_toggles/toggles.py:16]: > Should this be called something like InstanceTrackingMixin or InstanceListMixin?

I admit that `ListedTogglesMixin` is not a very good name. I'll go with `ClassInstancesMixin`, if you agree.

> This could be used by WaffleFlag, etc., but BaseToggle (or BaseSettingToggle) cannot.

Why not? I expect to inherit from this class once waffle flags have been migrated.

File: edx_toggles/toggles.py
Comment (regisb) [reply edx_toggles/toggles.py:18]: Actually, OrderedDict will sort the keys in the order they were inserted in the dictionary, but will not sort their values. Since we usually want to iterate on keys in alphabetical order (for instance to display them in the docs), it would make sense to use [sortedcontainer](https://pypi.org/project/sortedcontainers/).

File: edx_toggles/toggles.py
Comment (regisb) [reply edx_toggles/toggles.py:22]: You're right, I did not document this. I expect to add new feature toggles as follows:

```
MY_TOGGLE = SettingToggle.add("my_toggle", default=False, module_name=__name__)
```

And then, we don't need to pass the `module_name` when checking whether a feature toggle is enabled:

```
if SettingToggle("my_toggle", default=False).is_enabled():
    ...
```

Now that you mention it, `create` is probably more appropriate than `add`. I'll make that change.

File: edx_toggles/toggles.py
Comment (regisb) [reply edx_toggles/toggles.py:32]: Right! I referred to the ADR pull request instead of the actual document https://github.com/edx/edx-toggles/pull/56/files

File: edx_toggles/tests/test_toggles.py
Comment (regisb) [reply edx_toggles/tests/test_toggles.py:42]: :ok_hand:

File: edx_toggles/tests/test_toggles.py
Comment (regisb) [reply edx_toggles/tests/test_toggles.py:30]: I usually try to avoid unit tests that cover the same piece of code but ok :ok_hand:

File: Makefile
Comment (regisb) [reply Makefile:67]: I don't think so. I might be wrong, but I believe we need to run the Django tests in order to override the settings with the [`self.settings`](https://docs.djangoproject.com/en/dev/topics/testing/tools/#django.test.SimpleTestCase.settings) decorator.

File: edx_toggles/tests/test_toggles.py
Comment (regisb) [reply edx_toggles/tests/test_toggles.py:34]: :ok_hand:

File: edx_toggles/toggles.py
Comment (robrap) [reply edx_toggles/toggles.py:7]: Ok.  I was originally thinking that WaffleFlag and WaffleSwitch didn't have a default, but I guess they do have a default of False.

File: edx_toggles/toggles.py
Comment (robrap) [reply edx_toggles/toggles.py:32]: I understand now.  Let's just kill this comment and leave the `sorted` call below. If this ever becomes an issue, we can improve the implementation.

File: Makefile
Comment (robrap) [reply Makefile:67]: We use `pytest-django`.  Is that what usually makes this work?  See https://github.com/edx/edx-toggles/blob/master/requirements/test.in#L9. I don't know much about this config, but I would wonder how the cookie cutter works.  @jmbowman?

File: edx_toggles/toggles.py
Comment (robrap) [inline edx_toggles/toggles.py:36]: I now understand that this is a factory method.  But, I still don't see the value it adds, and would prefer to make things work like WaffleFlag.

You wrote:
>  ```
>  MY_TOGGLE = SettingToggle.add("my_toggle", default=False, module_name=__name__)
> ```
> And then, we don't need to pass the module_name when checking whether a feature toggle is enabled:

But, WaffleFlag doesn't need to pass `module_name` when checking if it is enabled.

Here is an example declaration and usage from edx-platform.  You can find them here: https://github.com/edx/edx-platform/search?q=HIDE_ACCESS_DENIALS_FLAG

**Example Declaration:**
```
HIDE_ACCESS_DENIALS_FLAG = WaffleFlag(
    waffle_namespace=COURSE_BLOCKS_API_NAMESPACE,
    flag_name=u'hide_access_denials',
    module_name=__name__,
)
```

**Example Usage:**
```
if HIDE_ACCESS_DENIALS_FLAG.is_enabled():
    ...
```

File: Makefile
Comment (regisb) [reply Makefile:67]: Right! I understand now. I'm not too familiar with pytest or tox. I'll update the tox.ini to get the tests to pass with `pytest`.

File: edx_toggles/toggles.py
Comment (robrap) [reply edx_toggles/toggles.py]: Thread moved to a new comment on the updated function: https://github.com/edx/edx-toggles/pull/66#discussion_r493019405

File: edx_toggles/toggles.py
Comment (robrap) [reply edx_toggles/toggles.py:34]: 1. I’d refactor that code to provide a constant for the instance.
2. The toggle state endpoint is also available on Open edX instances, and the module name on its own is included in the results For each toggle. Additionally, edX will use this to derive code_owner, but that is not its only purpose.
I still prefer it on the constructor.

File: edx_toggles/toggles.py
Comment (robrap) [reply edx_toggles/toggles.py]: Actually, it may make more sense for the user to sort of needed, or add sorted into the name. But, I don’t thing this is a big deal either way. I don’t imagine this will be used for other purposes.

File: edx_toggles/toggles.py
Comment (regisb) [reply edx_toggles/toggles.py]: :ok_hand: I'll get rid of the `create` method. As a consequence I'll also change the `_class_instances` attribute in a `WeakSet`, similar to the waffle_utils classes.

File: docs/decisions/0004-toggle-api.rst
Comment (robrap) [inline docs/decisions/0004-toggle-api.rst:26]: Thanks for this.  I was planning on a directory, rather than a module, because I imagine we want more modules.  Otherwise, an ``internal.py`` module would have been the simplest version of the idea.

```suggestion
Decision
--------

Objects that do not belong to the public API will be moved to an ``internal`` directory. Public API objects will be exposed as follows in ``edx_toggles/toggles/__init__.py``::

    from .internal.toggles import ...
    from .internal.waffle import ...

The benefits of this setup include:

* A clear designation of what is part of the public API.
* The ability to refactor the implementation without changing the API.
* A clear reminder to developers adding new code that it needs to be exposed if it is public.

Consequences
------------

Whenever a new class or function is added to the edx_toggles public API, it should be implemented in the ``internal`` directory and explicitly imported in the ``edx_toggles/toggles/__init__.py`` module.
```

**Review by robrap — COMMENTED** [review]
I'll need to come back to this to review the rest.

File: docs/decisions/0004-toggle-api.rst
Comment (regisb) [reply docs/decisions/0004-toggle-api.rst:26]: I purposefully used the word "module" and not "file" or "directory". This leaves us the possibility to split the `internal.py` file in multiple files later on. If you still feel like you prefer the wording with directories, I'll go with your change.

Comment (robrap) [reply docs/decisions/0004-toggle-api.rst:26]: 1. The example import you provided made me think "file", but if you are happy with "module", go for it.  What do you think of the example imports I provided, which I think is going to be our likely usage?
2. Re-reading the sentence `Objects that do not belong to the public API will be moved to..`, I find it potentially misleading. It reads as-if public methods could live outside of `internal`, which would in turn just be part of the public API.  As an alternative, 2 separate statements:
```
All implementation code should be moved to an ``internal`` module.
The Public API will be exposed as follows in ``edx_toggles/toggles/__init__.py``
```
3. I'd like to keep the `benefits` section I added.

**Review by robrap — APPROVED** [review]
@regisb: 
1. Does `WIP` in title represent anything I might be missing?
2. Please do your best to update the ADR according to the comments. Then, I guess you can merge. :) If I want to iterate on the ADR language, I can do that without blocking this PR.

File: docs/decisions/0004-toggle-api.rst
Comment (regisb) [reply docs/decisions/0004-toggle-api.rst:26]: :ok_hand:

**Comment by regisb** [comment]
> Does WIP in title represent anything I might be missing?

At the time it was a work-in-progress, and I forgot to remove the label once I considered this ready to merge.

> Please do your best to update the ADR according to the comments. Then, I guess you can merge. :) If I want to iterate on the ADR language, I can do that without blocking this PR.

Great, thanks!

**Pull request merged**
